### PR TITLE
Revive support for `URLClassLoader` in `ClassicPluginStrategy`

### PR DESCRIPTION
**Pull request by basil**
Once upon a time, Jenkins used to support both `URLClassLoader` and `AntClassLoader`, with `AntClassLoader` being preferred only because the `URLClassLoader` of the time didn't implement `Closeable`, which caused problems with predictable release of JAR file handles. That all changed in commit 47de54d070, which dropped support for `URLClassLoader` in order to introduce the Bytecode Compatibility Transformer. But the Bytecode Compatibility Transformer was removed in #5526, and `URLClassLoader` implements `Closeable` as of Java 7. So can we bring back support for `URLClassLoader`? Sure.

`ClassLoaderReflectionToolkit` needs to access internal `ClassLoader` methods for performance reasons. Kohsuke originally implemented this with reflection, but as of Java 11 this logs reflective access warnings, so Tim had to increase the visibility of some methods instead. The same approach can be done in a `URLClassLoader2` subclass of `URLClassLoader`, with a common `JenkinsClassLoader` interface for the methods required by `ClassLoaderReflectionToolkit`. I've done exactly that in this PR.

It all works great in both Java 11 and Java 8 (tested with `java -jar jenkins.war` and the 143 plugins on my staging server) modulo [this code](https://github.com/jenkinsci/bouncycastle-api-plugin/blob/58a04798909bd90b42a72b6d9a11acce5d0ee87f/src/main/java/jenkins/bouncycastle/api/BouncyCastlePlugin.java#L61-L66) in the BouncyCastle API plugin which needed to be patched to work with both `AntClassLoader` and `URLClassLoader2`. Because of the BouncyCastle API plugin issue, `URLClassLoader2` can't be enabled by default yet, so I am leaving `AntClassLoader` as the default for now. Once this PR makes it into an LTS release, I plan to follow up with a patch to BouncyCastle API to add support for `URLClassLoader2`. Then once that is merged, released, and widely adopted, we could make `URLClassLoader2` the default again. While that wouldn't completely eliminate our dependency on `AntClassLoader` (recall that `PluginFirstClassLoader` extends `AntClassLoader`), it would bring us far closer to the mainstream when it comes to class loading and therefore reduce our maintenance burden.

I plan to follow this up with a PR to update the [Jenkins Features Controlled with System Properties](https://www.jenkins.io/doc/book/managing/system-properties/) documentation.

### Proposed changelog entries

Internal: Experimental support for `URLClassLoader` can be enabled by setting `hudson.ClassicPluginStrategy.useAntClassLoader=false`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
+1 from me. It is a nice experimental feature to explore. It would be interesting to make the classloader even more pluggable via a String property, but it can be done later when needed. Thanks @basil !

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/jenkins/util/URLClassLoader2.java
Comment (jglick) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: Why not just define this in `JenkinsClassLoader`?

File: core/src/main/java/jenkins/util/JenkinsClassLoader.java
Comment (jglick) [inline core/src/main/java/jenkins/util/JenkinsClassLoader.java:22]: Why the `2`?

**Review by jglick — APPROVED** [review]
Anything that moves us towards being able to remove `AntClassLoader` sounds good to me.

**Review by jtnord — COMMENTED** [review]
please do not merge before my review

File: core/src/main/java/jenkins/util/JenkinsClassLoader.java
Comment (basil) [reply core/src/main/java/jenkins/util/JenkinsClassLoader.java:22]: Because [`ClassLoader#findLoadedClass`](https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#findLoadedClass-java.lang.String-) is final. (Also, I am not the one who introduced `findLoadedClass2`. Tim did in a previous PR.)

**Review by daniel-beck — APPROVED** [review]
The recent changes in `bouncycastle-api` are a bit of unfortunate timing 😦

File: core/src/main/java/jenkins/util/URLClassLoader2.java
Comment (jglick) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: You are adding a public `addURL` method in this PR, with the intention of allowing a revision of https://github.com/jenkinsci/bouncycastle-api-plugin/pull/46 to call it. You are also defining a `JenkinsClassLoader` to abstract the differences between `AntClassLoader` and `URLClassLoader2`. Rather than having that plugin have to go through a list of implementation classes looking for a method of this kind it could call, I am suggesting that the method be defined in one place, `JenkinsClassLoader`, with `AntClassLoader.addPathComponent` made into an implementation. So it seems in line with the goals of this PR; unless you want to _not_ do anything to help `bouncycastle-api` here, in which case there should be no need for a public `addURL` override either.

Comment (jglick) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: (This is a suggestion, not a blocking comment; I have already approved this PR.)

Comment (jtnord) [inline core/src/main/java/jenkins/util/URLClassLoader2.java:33]: this is restricted - but do we generically want to add arbitrary URLs to the classloader?
would it be better to take a `File` or `Path` and then convert it?

```suggestion
    @Override
    public void addJar(File jar) {
        super.addURL(jar.toURI().toURL());
    }
```

Comment (jtnord) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: > Today, the BouncyCastle API plugin calls a public method on a restricted AntClassLoader class, and my intention is to allow it to also call a public method on a restricted URLClassLoader2 class. This preserves the status quo, making things neither better nor worse.

Actually as I understand this, and this is my concern with the PR - it does make it slightly worse for users.
this PR as-is is not useful for the vast majority of users (if I read the Jenkins stats correctly almost every instance of Jenkins has the bouncycastle plugin installed - thus getting this confidence would take a long time) and by the time we enable it they would also have to have upgraded the BC plugin prior to that (so that is a lead time of 2 LTS bumps ~6 months) (and at least 3 months before users can opt in to test this _without branching the BC plugin_)
now bouncycastle is a detached plugin so we could update the version that is in the war, but if I recall correctly they do not get updated if already installed.

Would it not be better to put the code from `URLClassloader2` into the ANTClassloader, and the class that currently extends it can then take the code that is currently in `ANTClassloader`  that way - there is no need for a switch, the bouncycastle plugin continues to work, users do not need to ensure they have a compatible BC plugin when they upgrade a core that has the flag defaulted to true...

If we are not confident that we can use a subclass of `URLClassLoader` then what would make that possible?
If we are confident then the different approach would also seem to work as it is not changing the class hierarchy and should be binary compatible for everyone.

Happy to do an alternative PR to sketch this approach out.

**Review by jtnord — COMMENTED** [review]
Whilst I see no issues with the code, I am not convinced that this is the best approach for users.

there needs to be an update to the BC plugin that is installed before the core version can be changed.  as mentioned in the inline comments this is a bit of a chicken and egg and could cause breakage on upgrade (as the BC plugin is installed in a huge number of instances).

there is potentially a different solution I outlines that would seem to keep compatibility where it is needed, but has the downside of no system flag to disable it if things go wrong which I am happy to code up if we think this is worth perusing.

File: core/src/main/java/jenkins/util/URLClassLoader2.java
Comment (jglick) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: like `AntClassLoader.addPathComponent`

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jtnord) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:314]: if you are using something shared, then yes it will give potentially poorer results as obtaining the list can change between getting the size and then trying to construct.

Comment (basil) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:314]: Please read [Arrays of Wisdom of the Ancients](https://shipilev.net/blog/2016/arrays-wisdom-ancients/).

File: core/src/main/java/jenkins/util/URLClassLoader2.java
Comment (basil) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: As I wrote above, I do not intend to implement this RFE as part of this PR, though I consider it a reasonable idea.

Comment (timja) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: Not sure what this has to do with LTS?

Can't Bouncycastle API update to the weekly this is released in, and then add support for the `URLClassloader2` and release a new version?

and in the upgrade note when we get to LTS specify to make sure the plugin is upgraded?

Comment (jtnord) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: well a user on an LTS can not upgrade BC to a version with the changes, so you need to wait 2 LTS versions until you pull the plug (~ 5 months from now at the earliest). Basically LTS adds significant delays to be able to actually enable this flag (stats seems to imply > 90% of Jenkins installs have BC plugin installed) because you can not change the default until a user can install a version of the plugin that is compatible with core **before** they upgrade to that core (using the UI to manage plugin upgrades).

Sure we can branch the BC plugin - but how much testing would the weekly have? And if we think the URLClassloader is fine (I think we all do) then why do we even need to have something that breaks compatibility when I believe there is an approach that works regardless and requires no changes?

Comment (timja) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: I disagree about all this LTS talk, you don't need to wait, backports can be easily made if required, and BC doesn't appear to release very often anyway: https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk15on.

If you have a suggestion and are happy to provide an alternate PR then of course that will be looked at.

Comment (timja) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: We could also revert this change to bouncy castle that took this internal implementation detail and get a supported public API added first?

Comment (jtnord) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: > you don't need to wait, backports can be easily made if required

I think you have completely missed my point - you can not upgrade Jenkins and BC atomically.
I could release a BC targetting tomorrows weekly sure and in the next weekly we flip the flag.

How would an LTS user upgrade to the next .1 release if they have BC plugin installed without having a broken jenkins instance for some period of time?

Comment (jglick) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: `bouncycastle-api` is already using pretty awful hacks; at worst we could cut a release that had an older `jenkins.version` but worked with various implementation classes via reflection, then promptly push a new release with the new `jenkins.version` without the reflection.

(But still awful hacks. The root problem here IIUC is that BC itself is not designed correctly—you should be able to swap in a FIPS or standard implementation without replacing the API types.)

Comment (jtnord) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: you can not use BC-FIPS inside a plugin period.  the JVM needs to be configured and doing that in a plugin is too late as the JVM is up and configured, jetty has started already initialized some SSL context etc etc.

The API is mostly compatable (if you stay away from deprecated types).  not all providers are available in the regular BC (yup - there are some bits only in FIPS, when you expect the other way around).

we have been mostly able to swap a FIPS and non FIPS versions which is how the plugin can still work, but it is not possible to install a FIPS security provider in anything other than the java security config (or at best the first line of code in the `main` method.

**Comment by basil** [comment]
> it does make it slightly worse for users

No. This PR does not make anything worse for users in the slightest. The new experimental functionality introduced in this PR is off by default and currently unused (and unusable with BouncyCastle API 2.22 through 2.24).

With [e.g. commit 50ede7c](https://github.com/basil/bouncycastle-api-plugin/commit/50ede7c842df00fa648534f4f22f419fa9100ee1), which could be merged and released immediately (since it does not require a new baseline), users could begin to opt in to the new experimental functionality at the cost of an illegal reflective access in Java 11. Once BouncyCastle API uses a version of Jenkins core that contains this PR (or a common abstraction, as has been proposed above), BouncyCastle API could be improved to avoid the illegal reflective access by using `URLClassLoader2` (or the common abstraction).

At no point would we change the default class loader until we feel confident that the experimental functionality has been sufficiently tested. I am deliberately being extremely careful with PRs like this to ensure there is a graceful transition period without negative impact to users.

**Comment by jtnord** [comment]
> At no point would we change the default class loader until we feel confident that the experimental functionality has been sufficiently tested. I am deliberately being extremely careful with PRs like this to ensure there is a graceful transition period without negative impact to users.

this is not the intention I read from 

> can't be enabled by default yet, so I am leaving AntClassLoader as the default for now

That sounded like if it was not for BC then you would have made it the default pretty quickly.

Whilst THis PR does nothing the follow up one that makes it the default can and I am kind of treating them as one. because if you can not make the switch safely this impacts the whole of this PR

File: core/src/main/java/jenkins/util/URLClassLoader2.java
Comment (jtnord) [reply core/src/main/java/jenkins/util/URLClassLoader2.java:33]: > We could also revert this change to bouncy castle that took this internal implementation detail and get a supported public API added first?

so if the BouncyCastle API is going to use reflection to be compatible with this new feature I would not be overly happy doing that unless the new feature to be used (via reflection) was a public (aka not restricted) API - otherwise we will be in the same boat we are in today, but with reflection thrown into the mix for added fun.

Currently this is not compatible and adds no public API that can be migrated to.

It can not be reverted as it is actually required, but introducing reflection with a supported API is a path I would support if we are not going to go with an approach that maintains the compatability.

**Review by oleg-nenashev — APPROVED** [review]
After reading the conversation, I am re-iterating my support for this pull request. It is explicitly referenced as experimental, so any assumptions about URLClassLoader becoming a default soon are speculation at the moment. We will need to **eventually** figure out the BouncyCastle compatibility for the new mode, and the current analysis by @basil is a great start

Unless @jtnord explicitly requests changes, I suggest merging it into the next weekly release as **experimental** option.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Pull request merged**
